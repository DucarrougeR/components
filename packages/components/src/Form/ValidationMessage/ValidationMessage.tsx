import React, { forwardRef, Ref } from 'react'
import {
  CompatibleHTMLProps,
  CustomizableAttributes,
  layout,
  LayoutProps,
  reset,
  space,
  SpaceProps,
  typography,
  TypographyProps,
} from 'looker-design-tokens'
import styled from 'styled-components'

export type ValidationType = 'error'

export const CustomizableValidationMessageAttributes: CustomizableAttributes = {
  fontSize: 'xsmall',
}

export interface ValidationMessageProps
  extends LayoutProps,
    SpaceProps,
    TypographyProps,
    CompatibleHTMLProps<HTMLDivElement> {
  /**
   * The type of validation, therefore changing the message's text color. Accepts: error.
   */
  type?: ValidationType
  /**
   * The validation message to render.
   */
  message?: string
}

const ValidationMessageBase = styled.div.attrs(
  (props: ValidationMessageProps) => ({
    fontSize:
      props.fontSize || CustomizableValidationMessageAttributes.fontSize,
  })
)<ValidationMessageProps>`
  ${reset}
  ${props =>
    props.type === 'error'
      ? `color: ${props.theme.colors.palette.red700};`
      : ''}
  ${layout}
  ${space}
  ${typography}
`

export const ValidationMessage = forwardRef(
  ({ message, ...props }: ValidationMessageProps, ref: Ref<HTMLDivElement>) => (
    <ValidationMessageBase mr="xsmall" mt="xsmall" {...props} ref={ref}>
      {message}
    </ValidationMessageBase>
  )
)

ValidationMessage.displayName = 'ValidationMessage'
