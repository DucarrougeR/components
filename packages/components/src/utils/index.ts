export * from './GlobalStyle'
export * from './useToggle'
