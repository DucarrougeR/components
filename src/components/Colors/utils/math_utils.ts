/**
 * Utility to convert a numerical value to a percent given the value and a max value.
 */
export const toPerc = (value: number, maxValue: number) =>
  Math.round((value / maxValue) * 100)
