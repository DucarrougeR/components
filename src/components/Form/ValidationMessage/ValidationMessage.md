```js noeditor
const StatusAndResources = require('../../../../styleguide_components/StatusAndResources').StatusAndResources;

<StatusAndResources
  status="experimental"
  figmaURL="https://www.figma.com/file/2MG6DoSjk4IaLnjjFCnKFf/Forms"
  githubURL="https://github.com/looker/lens/blob/master/src/components/Form/ValidationMessage/ValidationMessage.tsx"
  feedbackTitle="ValidationMessage Component Feedback" />
```

### ValidationMessage

This component renders a validation message, and is used by Field components after they are validated.
