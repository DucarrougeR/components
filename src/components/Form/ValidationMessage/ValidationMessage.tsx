import * as React from 'react'
import { Box, BoxProps } from '../../Box'

export type ValidationType = 'error'

export interface ValidationMessageProps extends BoxProps {
  /**
   * The type of validation, therefore changing the message's text color. Accepts: error.
   */
  type?: ValidationType
  /**
   * The validation message to render.
   */
  message?: string
}

export const ValidationMessage: React.SFC<ValidationMessageProps> = ({
  type,
  message,
  ...props
}) => {
  const handleValidationType = () => {
    switch (type) {
      case 'error':
        return 'palette.red700'
      default:
        return undefined
    }
  }
  return (
    <Box
      {...props}
      mr="small"
      mt="small"
      fontSize="6"
      color={handleValidationType()}
    >
      {message}
    </Box>
  )
}
