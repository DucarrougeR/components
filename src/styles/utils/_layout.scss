// Returns a value from the $spacing list
@function spacing($key) {
  @return map-lookup($spacing, $key, '$spacing');
}
