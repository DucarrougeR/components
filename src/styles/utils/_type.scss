// Returns a value from the $test-size list
@function font-size($key) {
  @return map-lookup($text-size, $key, '$text-size');
 }

 // Returns a value from the $line-height list
 @function line-height($key) {
   @return map-lookup($line-height, $key, '$line-height');
 }

 @mixin fontSize($size) {
  font-size: $size;
  font-size: pxToRem($size);
}
