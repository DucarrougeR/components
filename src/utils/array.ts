/**
 * Returns an array of length `size` filled with zeros.
 * Useful for looping constructs where a length is known, but no array is provided.
 */

export const sizedArray = (size: number) => Array(size).fill(0)
