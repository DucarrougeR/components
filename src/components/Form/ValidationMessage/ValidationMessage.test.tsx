import * as React from 'react'
import { createWithTheme } from '../../../../test/utils/create_with_theme'
import { ValidationMessage } from './ValidationMessage'

test('An error message', () => {
  const component = createWithTheme(
    <ValidationMessage type="error">Error!</ValidationMessage>
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A warning message', () => {
  const component = createWithTheme(
    <ValidationMessage type="warning">Warning</ValidationMessage>
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('An info message', () => {
  const component = createWithTheme(
    <ValidationMessage type="info">Info</ValidationMessage>
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A success message', () => {
  const component = createWithTheme(
    <ValidationMessage type="success">Success!</ValidationMessage>
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
