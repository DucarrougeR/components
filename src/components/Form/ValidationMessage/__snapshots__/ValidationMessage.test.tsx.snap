// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`An error message 1`] = `
.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  color: #990F14;
  font-size: 0.75rem;
  margin-right: 0.5rem;
  margin-top: 0.5rem;
}

<div
  className="c0"
>
  Error!
</div>
`;
